src/helpers/buffers/unordered_receiver.rs

```diff
@@ -204,29 +209,54 @@ where
     /// Poll for the next record.  This should only be invoked when
     /// the future for the next message is polled.
     fn poll_next<M: Message>(&mut self, cx: &mut Context<'_>) -> Poll<Result<M, Error>> {
-        if let Some(m) = self.spare.read() {
-            self.wake_next();
-            return Poll::Ready(Ok(m));
-        }
-
-        loop {
-            match self.stream.as_mut().poll_next(cx) {
-                Poll::Pending => {
-                    return Poll::Pending;
-                }
-                Poll::Ready(Some(b)) => {
-                    if let Some(m) = self.spare.extend(b.as_ref()) {
-                        self.wake_next();
-                        return Poll::Ready(Ok(m));
+        // If spare has enough data for us, poll it first
+        // otherwise, poll the underlying stream until it returns pending or it provides enough
+        // data to return a value.
+
+        let message = self.spare.read::<M>();
+        let next = match message {
+            Some(m) => {
+                // this check exists to make sure the inner stream is eventually moved to
+                // the closed state. We don't want to poll it too often, but we also need to know
+                // when it is done and `UnorderedReceiver` can be dropped.
+                if self.spare.is_empty() {
+                    // we don't want to be woken up here, control loop is driven by the client.
+                    // They decide when they want the next message and must issue a `poll` for it.
+
+                    // TODO: https://github.com/rust-lang/rust/issues/98286
+                    let mut cx = Context::from_waker(futures::task::noop_waker_ref());
+                    match self.stream.as_mut().poll_next(&mut cx) {
```

Comment:
Does this even need to poll the stream if it is using spare bytes?  The next call will then hit the `None` arm of the outer match, which polls the stream properly.

I don't think that it is a good idea to read proactively like this.  We've already got what we need, so this will only inflate our buffer and fail to provide proper back pressure.  Given that we're waking the next item as noted, I don't see why the next poll won't work.